### PR TITLE
WIP: optionally configure DirectDruidClient to use /druid/v3 instead of /druid/v2

### PR DESCRIPTION
**Pull request by himanshug**
for talking to historicals and accumulate the response context.

DEPENDS ON #3319 and is reviewable after same is merged , creating PR for lookahead

enabled with "druid.broker.httpClient.useV3QueryUrl=true" in runtime.properties.

response context from broker would include response context from various historicals as...

{
    "result": [.... ]
    "context": {
        .....
        "historical1" : {...},
        "historical2" : {....}
    }

### CODE REVIEW

**Comment by drcrallen** [comment]
Proposing as part of 0.9.3 for now as not to block 0.9.2

File: server/src/main/java/io/druid/client/BrokerServerView.java [hunk 1]
```diff
@@ -200,7 +203,15 @@ private QueryableDruidServer addServer(DruidServer server)
...
   {
-    return new DirectDruidClient(warehouse, queryWatcher, smileMapper, httpClient, server.getHost(), emitter);
+    return new DirectDruidClient(
+        warehouse,
+        queryWatcher,
+        smileMapper,
+        httpClient,
+        server.getHost(),
+        emitter,
+        directDruidClientConfig.isUseV3QueryUrl()
```
Comment (drcrallen) [inline server/src/main/java/io/druid/client/BrokerServerView.java]: can this just be a string directly which defaults to "v2"?

File: server/src/main/java/io/druid/client/DirectDruidClient.java [hunk 2]
```diff
@@ -386,22 +389,42 @@ public void onFailure(Throwable t)
       throw Throwables.propagate(e);
     }
 
-    Sequence<T> retVal = new BaseSequence<>(
-        new BaseSequence.IteratorMaker<T, JsonParserIterator<T>>()
-        {
-          @Override
-          public JsonParserIterator<T> make()
+    Sequence<T> retVal;
+    if (useV3QueryUrl) {
```
Comment (drcrallen) [inline server/src/main/java/io/druid/client/DirectDruidClient.java]: I propose using a factory here with the two default implementations being a queryv2 and a queryv3 factory which can generate the proper return value.

File: server/src/main/java/io/druid/client/DirectDruidClient.java [hunk 3]
```diff
@@ -509,4 +532,138 @@ public void close() throws IOException
...
+          throw new QueryInterruptedException(e, host);
+        }
+      }
+    }
+
+    @Override
+    public void close() throws IOException
+    {
+      if (jp != null) {
+        jp.close();
```
Comment (drcrallen) [inline server/src/main/java/io/druid/client/DirectDruidClient.java]: suggest adding `jp = null;` after this.

File: server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java [hunk 4]
```diff
@@ -161,9 +164,9 @@ protected void service(HttpServletRequest request, HttpServletResponse response)
     final String defaultHost = hostFinder.getDefaultHost();
     request.setAttribute(HOST_ATTRIBUTE, defaultHost);
 
-    final boolean isQueryEndpoint = request.getRequestURI().startsWith("/druid/v2");
+    final String requestURI = request.getRequestURI();
 
-    if (isQueryEndpoint && HttpMethod.DELETE.is(request.getMethod())) {
+    if (requestURI.startsWith(QUERY_URL_V2) && HttpMethod.DELETE.is(request.getMethod())) {
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java]: This kind of logic strikes me as something that should be able to be handled at a more abstract routing level.  Have you had a chance to look for such items?

**Comment by fjy** [comment]
@himanshug is this still WIP?

**Comment by himanshug** [comment]
@fjy it was ready bu depends on #3319 so removing from 0.9.3 till that is merged. also i'm planning to make additional changes to #3319 which will require changes here too. see https://github.com/druid-io/druid/pull/3319#issuecomment-259475020

**Pull request closed without merging**